### DISCUSSION TITLE
Tasmota ESP32-CAM: need help with rule to check for night

### DISCUSSION THREAD

**Discussion by mrneutron42** [arendst/Tasmota#16378] [Q&A]
I'm running pre-compiled Tasmota 12.0.2 for ESP32-CAM.  I have been looking for examples of simple rules to detect night in
https://tasmota.github.io/docs/Commands/#rules

In the Cookbook section I found a rule that is close to what I want, but it seems overly complicated and takes 2 rules to accomplish.
https://tasmota.github.io/docs/Rules/#turn-on-light-before-dawn-and-at-dusk

The rules use "event chksun" which is a command I cannot find rules list.
Is it's use and behavior defined somewhere?
I've tried to copy it and simplify the 2 rules into one rule that checks the time to see if it is night - a time between sunset and sunrise.  If its night, do the command "wcagc 1", which is a command specific to an ESP32-CAM, that turns on the automatic gain control.

Rule2
  ON Time#Initialized DO backlog wcagc 0; event chksun ENDON
  ON backlog event chksunrise=%time%; event chksunset=%time% ENDON
  ON event#chksunrise<%sunrise% DO wcagc 1 ENDON
  ON event#chkSunset>%sunset% DO wcagc 1 ENDON

That didn't work. wcagc is set to 0 all the time.

A multiple condition IF statement seems like it would work better.
Will the IF AND construction work with the precompiled version of Tasmota ESP32-CAM?

Rule2
  ON Time#Initialized DO backlog wcagc 0; event chksun ENDON
  ON backlog event chksunrise=%time%; event chksunset=%time% ENDON
  ON Time#Minute IF (event#chksunrise<%sunrise%) {AND} (event#chksunset>%sunset%) DO wcagc 1 ENDIF ENDON

**Comment by sfromis** [comment]
Instead of all that, as a starting point the simplest strategy is to use "Configure Timers" to get rule events for times relative to sunrise/sunset, as long as the device is stable.

The `Event` command is documented on the page for commands. `chksun` is just a name you select for the event.                                                                                                                                

>  ON backlog event chksunrise=%time%; event chksunset=%time% ENDON

> ON Time#Minute IF (event#chksunrise<%sunrise%) {AND} (event#chksunset>%sunset%) 

This is "very wrong". you cannot sprinkle trigger expressions with `#` anywhere you want, only between `on` and `do`. and those `{}` are not part of the syntax for `AND`. And `Backlog is a command, which is misplaced in the trigger part of a rule. Back to square one for rules syntax, but:

In general, such mixes of events and logic is tricky to get right in rules, and I'd really recommend ditching rules for what you try to do. You have much better control with Berry code instead.

**Reply by mrneutron42** [reply]
I see Berry scripting is only available in ESP32 builds of Tasmota.  
Most of the devices I have use standard Tasmota, so learning rules is still a worthwhile long term pursuit.  
Can you suggest edits to the rule tries I posted?
I have been reading the Rules documentation, but it is not very clear to me.  
I need to see more concrete examples of working rules, so I can better see the usage patterns and copy the patterns.
Thanks.

**Reply by sfromis** [reply]
Yes, and ESP32-CAM is ESP32. With Berry you have much better features.
If you want to learn to do tricky stuff in rules, start small and play around. Many concrete examples of rules on the docs page, cookbook section.

**Comment by mrneutron42** [comment]
I looked closer at the Timer configuration page and it does show "Rule" as an action.
It simply says "Rule", not "Rule 1" or Rule 2", etc.
Which Rule will it trigger?
![timer rule number](https://user-images.githubusercontent.com/70338169/188670029-7b5758ce-005a-4d30-a3c9-30d7e941f1c2.JPG)

**Reply by sfromis** [reply]
It will trigger the rule with a trigger to match the timer number. Rules are not numbered, and you can have many rules in each of the 3 rule sets.

**Reply by arendst** [reply]
All. It merely says it will execute the rule processing where you added a trigger containing the timer.

**Reply by mrneutron42** [reply]
OK, I think I understand.
I just made 2 timer events, 1 for sunrise and 2 for sunset.
and 2 rules, 1 for sunrise and 2 for sunset.
Rule2 ON Clock#Timer=1 backlog wcagc 1; wcagcgain 0; wcgainceiling 0 ENDON
Rule2 1
Rule3 ON Clock#Timer=2 backlog wcagc 1; wcagcgain 3; wcgainceiling 3 ENDON
Rule3 1
![timer1 for sunrise](https://user-images.githubusercontent.com/70338169/188674613-673417d3-6698-44a9-a667-c02394c7d2fa.JPG)

**Reply by barbudor** [reply]
Almost
Your rules are lacking the 'DO' keyword

And to re-emphase what sfromis said, you don't need to use 2 rules set. You can have both rules inside Rule2  or Rule3

**Reply by mrneutron42** [reply]
Oh.  I figured there was some limit to the amount of content a rule could contain.
How is this?
Rule2 ON Clock#Timer=1 DO backlog wcagc 1; wcagcgain 0; wcgainceiling 0 ENDON ON Clock#Timer=2 DO backlog wcagc 1; wcagcgain 3; wcgainceiling 3 ENDON

**Reply by barbudor** [reply]
Looks like ok

There is indeed a limitation of 512 bytes per rule-set. But as it can use compression this could represent 2000 or more characters
